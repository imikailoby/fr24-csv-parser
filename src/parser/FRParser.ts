import { parseCsv, validateCsvColumns } from './csvParser';
import { parseCsvRow } from './rowParser';
import { generateResultObject } from '../utils/generators';
import type { FRData } from '../types/models';
import { CsvDataRowIndex, EXPECTED_CSV_COLUMNS } from '../constants/csv';

export class FRParser {
  /** Process the CSV content and return formatted data. */
  public parse(csv: string): FRData[] {
    validateCsvColumns(csv, EXPECTED_CSV_COLUMNS);
    const parsedData = parseCsv(csv);
    const bodyRows = parsedData.slice(CsvDataRowIndex.FIRST_BODY_ROW);

    const result = this.convert(bodyRows);
    return result;
  }

  private convert(bodyRows: string[][]): FRData[] {
    return bodyRows.map((row) => {
      const parsedCsv = parseCsvRow(row);
      return generateResultObject(parsedCsv);
    });
  }
}
