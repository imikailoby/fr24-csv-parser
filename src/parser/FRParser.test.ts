import { mockedCsvContent } from '../tests/mocks';
import { FRParser } from './FRParser';

describe('FRParser', () => {
  it('parse', () => {
    expect(new FRParser().parse(mockedCsvContent)).toMatchSnapshot('parse-default');
  });
});
