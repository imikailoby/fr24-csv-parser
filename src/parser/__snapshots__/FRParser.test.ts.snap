// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FRParser parse: parse-default 1`] = `
[
  {
    "aircraft": {
      "code": "A339",
      "id": "2130",
      "manufacturer": "Airbus",
      "model": "A330-900neo",
      "registration": null,
    },
    "airline": {
      "iata": "DL",
      "icao": "DAL",
      "id": "223",
      "name": "Delta Air Lines",
    },
    "flight": {
      "arrivalId": "1328",
      "arrivalTime": "14:30:00",
      "date": "2024-10-12",
      "departureId": "122",
      "departureTime": "12:10:00",
      "duration": "08:20:00",
      "number": "DL49",
    },
    "from": {
      "airport": "Schiphol",
      "city": "Amsterdam",
      "iata": "AMS",
      "icao": "EHAM",
    },
    "passenger": {
      "cabin": "economy",
      "note": null,
      "reason": "leisure",
      "seat": {
        "number": "1A",
        "type": "window",
      },
    },
    "to": {
      "airport": "John F Kennedy",
      "city": "New York",
      "iata": "JFK",
      "icao": "KJFK",
    },
  },
]
`;
